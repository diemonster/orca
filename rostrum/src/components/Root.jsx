import React from 'react';
import { Route, Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import configureStore from '../store/configureStore';
import History from '../middleware/history';

import AuthClient from '../middleware/authClient';
import K8sClient from '../middleware/k8sClient';
import { initialState as authInitialState } from '../reducers/authReducer';
import { initialState as configInitialState } from '../reducers/configReducer';
import { initialState as k8sInitialState } from '../reducers/k8sReducer';

import App from './App';


class Root extends React.Component {
  componentWillMount() {
    // by spreading both the initial state and another object, we make
    // sure that the initial state is fully initialized and then allow
    // for any values that might be passed in externally to overwrite
    // keys as needed
    const config = { ...configInitialState, ...window.config };

    const authClient = (
      config.authClient
        ? config.authClient
        : new AuthClient(config.auth0Domain, config.auth0ClientID)
    );

    const auth = { ...authInitialState, authClient };

    const k8sClient = (
      config.k8sClient
        ? config.k8sClient
        : new K8sClient(config.auth0Domain, config.auth0ClientID)
    );

    const k8s = { ...k8sInitialState, k8sClient };

    const initialState = { auth, config, k8s };
    this.store = configureStore(initialState);
  }

  render() {
    return (
      <Provider store={this.store}>
        <Router history={History}>
          <Route component={App} />
        </Router>
      </Provider>
    );
  }
}

export default Root;
