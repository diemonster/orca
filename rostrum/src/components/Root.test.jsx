import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Provider } from 'react-redux';
import { Router, Switch } from 'react-router-dom';
import Root from './Root';
import Authenticator from '../authenticator/authenticator';
import History from '../authenticator/history';

Enzyme.configure({ adapter: new Adapter() });

describe('components', () => {
  describe('Root', () => {
    beforeEach(() => {
      window.config = {
        auth0Domain: 'test.com',
        auth0ClientID: '1234567890',
      };
    });

    it('should render itself and its subcomponents', () => {
      const expectedState = {
        config: {
          auth0Domain: 'test.com',
          auth0ClientID: '1234567890',
        },
        namespace: {
          namespaceCreateInput: '',
          namespaceDeleteInput: '',
          namespaceObjects: [],
        },
        rolebinding: {
          namespacedRolebindings: {},
        },
      };

      const root = shallow(<Root />);

      expect(root).toMatchSnapshot();
      expect(root.find(Provider).prop('store').getState()).toEqual(expectedState);
      expect(root.find(Router).prop('history')).toEqual(History);
      expect(root.find(Switch).children().length).toEqual(2);
      expect(root.find(Switch).childAt(0).prop('path')).toEqual('/callback');
    });

    it('should set properties on componentWillMount', () => {
      const expectedState = {
        config: {
          auth0Domain: 'test.com',
          auth0ClientID: '1234567890',
        },
        namespace: {
          namespaceCreateInput: '',
          namespaceDeleteInput: '',
          namespaceObjects: [],
        },
        rolebinding: {
          namespacedRolebindings: {},
        },
      };

      const instance = shallow(<Root />).instance();

      const authenticator = new Authenticator(
        window.config.auth0Domain,
        window.config.auth0ClientID,
      );

      authenticator.getProfile = (cb) => {
        this.auth0.client.userInfo('', cb);
      };

      expect(instance.config).toEqual(window.config);
      expect(instance.store.getState()).toEqual(expectedState);
      // I don't know if there's a better way to check for equality of two
      // discrete but identical instances. Using JSON.stringify() here returns an
      // error about converting a circular structure, and
      // expect(instance.authenticator).toEqual(authenticator) returns false
      // even though jest states that there is no visible diff.
      expect(`${instance.authenticator}`).toEqual(`${authenticator}`);
    });

    // it('should generate the right <Auth> component', () => {
    // const instance = shallow(<Root />).instance();
    // instance.authenticator.getProfile = jest.fn();
    // const auth = shallow(instance.authWithProps());
    //
    // expect(auth.prop('authenticator')).toBe(instance.authenticator);
    // });
  });
});
