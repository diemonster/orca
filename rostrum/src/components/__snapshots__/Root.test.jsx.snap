// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Root should render itself and its subcomponents 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <Route
        exact={true}
        path="/callback"
        render={[Function]}
      />
      <Route
        render={[Function]}
      />
    </Switch>
  </Router>
</Provider>
`;
